File: benchmarks/medium/M2-csv-processing/csharp/Program.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Security.Cryptography;
using System.Text;

CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;

const int N = 1_000_000;
const int RUNS = 10;
const string CsvFile = "input.csv";
const string OutputFile = "output.csv";

string[] firstNames = ["James", "Mary", "John", "Patricia", "Robert",
                       "Jennifer", "Michael", "Linda", "David", "Elizabeth"];
string[] lastNames = ["Smith", "Johnson", "Williams", "Brown", "Jones",
                      "Garcia", "Miller", "Davis", "Rodriguez", "Martinez"];
string[] departments = ["Engineering", "Marketing", "Sales", "HR",
                        "Finance", "Operations", "Support", "Legal"];

Console.WriteLine("=== BENCHMARK: M2 - CSV Processing ===");
Console.WriteLine("Language: C#");
Console.WriteLine($"Runtime:  .NET {Environment.Version}");
Console.WriteLine();

GenerateCsv();

double[] timings = new double[RUNS];
Stopwatch sw = new();

for (int run = 0; run < RUNS; run++)
{
    sw.Restart();
    Process();
    sw.Stop();
    timings[run] = sw.Elapsed.TotalMilliseconds;
    Console.WriteLine($"Run {run + 1,2}: {timings[run]:F2} ms");
}

Console.WriteLine();
double min = timings[0], max = timings[0], sum = 0;
foreach (double t in timings)
{
    sum += t;
    if (t < min) min = t;
    if (t > max) max = t;
}
double avg = sum / RUNS;
double variance = 0;
foreach (double t in timings)
    variance += (t - avg) * (t - avg);
variance /= RUNS;
double stddev = Math.Sqrt(variance);

Console.WriteLine($"Min:    {min:F2} ms");
Console.WriteLine($"Avg:    {avg:F2} ms");
Console.WriteLine($"Max:    {max:F2} ms");
Console.WriteLine($"StdDev: {stddev:F2} ms");
Console.WriteLine();

// Verification: SHA-256 of output.csv
byte[] outputBytes = File.ReadAllBytes(OutputFile);
byte[] hashBytes = SHA256.HashData(outputBytes);
Console.WriteLine($"Verification: SHA-256 = {Convert.ToHexString(hashBytes).ToLowerInvariant()}");

File.Delete(CsvFile);
File.Delete(OutputFile);

void GenerateCsv()
{
    using var writer = new StreamWriter(CsvFile, false, new UTF8Encoding(false), 1 << 20);
    writer.NewLine = "\n";
    writer.WriteLine("id,first_name,last_name,email,department,salary");
    uint state = 42u;
    for (int i = 0; i < N; i++)
    {
        state = unchecked(state * 1664525u + 1013904223u);
        double salary = 30000.0 + (state / 4294967295.0) * 120000.0;
        string first = firstNames[i % 10];
        string last = lastNames[i % 10];
        string dept = departments[i % 8];
        writer.WriteLine($"{i + 1},{first},{last},{first.ToLowerInvariant()}.{last.ToLowerInvariant()}@company.com,{dept},{salary:F2}");
    }
}

void Process()
{
    var groups = new Dictionary<string, (int count, double total, double max, double min)>(8);

    using (var reader = new StreamReader(CsvFile))
    {
        reader.ReadLine(); // skip header
        string? line;
        while ((line = reader.ReadLine()) != null)
        {
            // Parse salary (last field) and department (second to last)
            int lastComma = line.LastIndexOf(',');
            double salary = double.Parse(line.AsSpan(lastComma + 1));
            if (salary > 75000.0)
            {
                int prevComma = line.LastIndexOf(',', lastComma - 1);
                string dept = line.Substring(prevComma + 1, lastComma - prevComma - 1);
                if (groups.TryGetValue(dept, out var g))
                {
                    groups[dept] = (g.count + 1, g.total + salary, Math.Max(g.max, salary), Math.Min(g.min, salary));
                }
                else
                {
                    groups[dept] = (1, salary, salary, salary);
                }
            }
        }
    }

    var results = new List<(string dept, int count, double avg, double max, double min)>(groups.Count);
    foreach (var (dept, g) in groups)
        results.Add((dept, g.count, g.total / g.count, g.max, g.min));

    results.Sort((a, b) => b.avg.CompareTo(a.avg));

    using var writer = new StreamWriter(OutputFile, false, new UTF8Encoding(false));
    writer.NewLine = "\n";
    writer.WriteLine("department,count,avg_salary,max_salary,min_salary");
    foreach (var (dept, count, avgSal, maxSal, minSal) in results)
        writer.WriteLine($"{dept},{count},{avgSal:F2},{maxSal:F2},{minSal:F2}");
}

File: benchmarks/medium/M2-csv-processing/fsharp/Program.fs
open System
open System.Collections.Generic
open System.Diagnostics
open System.Globalization
open System.IO
open System.Security.Cryptography
open System.Text

[<EntryPoint>]
let main _ =
    CultureInfo.CurrentCulture <- CultureInfo.InvariantCulture
    let n = 1_000_000
    let runs = 10
    let csvFile = "input.csv"
    let outputFile = "output.csv"

    let firstNames = [| "James"; "Mary"; "John"; "Patricia"; "Robert";
                        "Jennifer"; "Michael"; "Linda"; "David"; "Elizabeth" |]
    let lastNames = [| "Smith"; "Johnson"; "Williams"; "Brown"; "Jones";
                       "Garcia"; "Miller"; "Davis"; "Rodriguez"; "Martinez" |]
    let departments = [| "Engineering"; "Marketing"; "Sales"; "HR";
                         "Finance"; "Operations"; "Support"; "Legal" |]

    printfn "=== BENCHMARK: M2 - CSV Processing ==="
    printfn "Language: F#"
    printfn "Runtime:  .NET %O" Environment.Version
    printfn ""

    // Generate CSV
    do
        use writer = new StreamWriter(csvFile, false, UTF8Encoding(false), 1 <<< 20)
        writer.NewLine <- "\n"
        writer.WriteLine("id,first_name,last_name,email,department,salary")
        let mutable state = 42u
        for i in 0 .. n - 1 do
            state <- state * 1664525u + 1013904223u
            let salary = 30000.0 + (float state / 4294967295.0) * 120000.0
            let first = firstNames.[i % 10]
            let last = lastNames.[i % 10]
            let dept = departments.[i % 8]
            writer.WriteLine($"{i + 1},{first},{last},{first.ToLowerInvariant()}.{last.ToLowerInvariant()}@company.com,{dept},{salary:F2}")

    let timings = Array.zeroCreate<float> runs
    let sw = Stopwatch()

    let processFile () =
        let groups = Dictionary<string, struct (int * float * float * float)>(8)

        use reader = new StreamReader(csvFile)
        reader.ReadLine() |> ignore // skip header
        let mutable line = reader.ReadLine()
        while not (isNull line) do
            let lastComma = line.LastIndexOf(',')
            let salary = Double.Parse(line.AsSpan(lastComma + 1))
            if salary > 75000.0 then
                let prevComma = line.LastIndexOf(',', lastComma - 1)
                let dept = line.Substring(prevComma + 1, lastComma - prevComma - 1)
                match groups.TryGetValue(dept) with
                | true, struct (c, t, mx, mn) ->
                    groups.[dept] <- struct (c + 1, t + salary, max mx salary, min mn salary)
                | false, _ ->
                    groups.[dept] <- struct (1, salary, salary, salary)
            line <- reader.ReadLine()

        let results = ResizeArray(groups.Count)
        for kv in groups do
            let struct (c, t, mx, mn) = kv.Value
            results.Add(struct (kv.Key, c, t / float c, mx, mn))

        results.Sort(fun (struct (_, _, a, _, _)) (struct (_, _, b, _, _)) -> compare b a)

        use writer = new StreamWriter(outputFile, false, UTF8Encoding(false))
        writer.NewLine <- "\n"
        writer.WriteLine("department,count,avg_salary,max_salary,min_salary")
        for struct (dept, count, avgSal, maxSal, minSal) in results do
            writer.WriteLine($"{dept},{count},{avgSal:F2},{maxSal:F2},{minSal:F2}")

    for run in 0 .. runs - 1 do
        sw.Restart()
        processFile ()
        sw.Stop()
        timings.[run] <- sw.Elapsed.TotalMilliseconds
        printfn "Run %2d: %.2f ms" (run + 1) timings.[run]

    printfn ""
    let minT = Array.min timings
    let maxT = Array.max timings
    let avg = Array.average timings
    let variance = timings |> Array.averageBy (fun t -> (t - avg) ** 2.0)
    let stddev = sqrt variance

    printfn "Min:    %.2f ms" minT
    printfn "Avg:    %.2f ms" avg
    printfn "Max:    %.2f ms" maxT
    printfn "StdDev: %.2f ms" stddev
    printfn ""

    // Verification: SHA-256 of output.csv
    let outputBytes = File.ReadAllBytes(outputFile)
    let hashBytes = SHA256.HashData(outputBytes)
    printfn "Verification: SHA-256 = %s" (Convert.ToHexString(hashBytes).ToLowerInvariant())

    File.Delete(csvFile)
    File.Delete(outputFile)
    0
